modules/stochastic_tools/src/utils/CovarianceFunction.C

```diff
@@ -0,0 +1,402 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CovarianceFunction.h"
+#include "GaussianProcessTrainer.h"
+#include "MooseError.h"
+#include "DataIO.h"
+#include "libmesh/auto_ptr.h"
+
+#include "libmesh/dense_matrix_impl.h"
+
+#include <cmath>
+
+namespace CovarianceFunction
+{
+
+MooseEnum
+makeCovarianceKernelEnum()
+{
+  return MooseEnum("squared_exponential=0 exponential=1 matern=2 matern_half_int=3");
+}
+
+std::unique_ptr<CovarianceKernel>
+makeCovarianceKernel(const MooseEnum & kernel_type, const GaussianProcessTrainer * gp)
+{
+
+  std::vector<Real> length_factor = gp->getParam<std::vector<Real>>("length_factor");
+  Real signal_variance = gp->getParam<Real>("signal_variance");
+  Real noise_variance = gp->getParam<Real>("noise_variance");
+
+  if (kernel_type == 0)
+  {
+    std::unique_ptr<CovarianceKernel> ptr = libmesh_make_unique<SquaredExponential>();
+    ptr->set_signal_variance(signal_variance);
+    ptr->set_noise_variance(noise_variance);
+    ptr->set_length_factor(length_factor);
```

Comment:
Can these be arguments for the classes' constructor? I don't really like these set functions. We could have some spurious errors if things are not done in the right order.

Reply:
I could move the initial setting of these to the constructor, but these functions will need to exist. In the near future when the automatic parameter optimization comes in these mutators will be called to update the hyper-parameters.

Reply:
For this case, I think we have the member variables be references to the arguments. That way when we change a parameter in the Trainer, these variables will be automatically updated. I probably need to think about this more though.

Reply:
I think what @zachmprince is suggesting is that the items that feed into the `CovarianceKernel` object should be what is declared as model data. Then the `CovarianceKernel` in trainer and/or model can operate from the same data, so if the trainer changes it then the model will automatically have the data.

The `CovarianceKernel` will look something like this:

```
class CK
{
public:
  CK(const Real & foo), _foo(foo){}

  private:
   const Real & _foo;
}
```

Reply:
Where the foo passed into this class is the reference returned from the declareModelData<Real>("foo")

Reply:
Doing this should also remove the need to have the dataStore/Load functions for the classes, since you would just instantiate the class using the loaded data.

Reply:
This is now all taken care of by the CovarianceFunction MooseObjects